File: mp3_convert.py
# mp3_convert.py
import subprocess
import argparse
import os
import sys
import shutil
from pathlib import Path

SUPPORTED_INPUT = ('.wav', '.flac', '.ogg', '.m4a', '.aac', '.wma', '.aiff', '.mp3', '.m4b')
SUPPORTED_OUTPUT = ('mp3', 'wav', 'flac', 'ogg', 'm4a', 'aac')

def check_ffmpeg():
    """Проверяет наличие ffmpeg в PATH."""
    if shutil.which("ffmpeg") is None:
        print("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.", file=sys.stderr)
        sys.exit(1)

def build_ffmpeg_cmd(input_file, output_file, bitrate, channels, sample_rate, normalize, format_out):
    """Формирует команду ffmpeg с учётом всех параметров."""
    cmd = ["ffmpeg", "-i", input_file, "-y"]
    
    # Битрейт
    if bitrate:
        cmd.extend(["-b:a", f"{bitrate}k"])
    
    # Частота дискретизации
    if sample_rate:
        cmd.extend(["-ar", str(sample_rate)])
    
    # Количество каналов
    if channels:
        cmd.extend(["-ac", str(channels)])
    
    # Нормализация громкости (EBU R128)
    if normalize:
        cmd.extend(["-af", "loudnorm=I=-16:LRA=11:TP=-1.5"])
    
    # Кодек (MP3 по умолчанию)
    if format_out == 'mp3':
        cmd.extend(["-c:a", "libmp3lame"])
    elif format_out == 'wav':
        cmd.extend(["-c:a", "pcm_s16le"])
    elif format_out == 'flac':
        cmd.extend(["-c:a", "flac"])
    elif format_out == 'ogg':
        cmd.extend(["-c:a", "libvorbis"])
    elif format_out == 'm4a' or format_out == 'aac':
        cmd.extend(["-c:a", "aac"])
    
    cmd.append(output_file)
    return cmd

def convert_file(input_path, output_path, bitrate, channels, sample_rate, normalize, format_out):
    """Конвертирует один файл."""
    cmd = build_ffmpeg_cmd(input_path, output_path, bitrate, channels, sample_rate, normalize, format_out)
    print(f"  Выполняется: {' '.join(cmd)}")
    try:
        subprocess.run(cmd, check=True, capture_output=True)
        return True
    except subprocess.CalledProcessError as e:
        print(f"  Ошибка: {e.stderr.decode()}", file=sys.stderr)
        return False

def find_audio_files(root, recursive):
    """Находит все аудиофайлы в указанной директории."""
    files = []
    root_path = Path(root)
    if root_path.is_file() and root_path.suffix.lower() in SUPPORTED_INPUT:
        return [root_path]
    if not root_path.is_dir():
        return []
    pattern = "**/*" if recursive else "*"
    for p in root_path.glob(pattern):
        if p.is_file() and p.suffix.lower() in SUPPORTED_INPUT:
            files.append(p)
    return files

def main():
    parser = argparse.ArgumentParser(description='Конвертер аудио в MP3 (и обратно)')
    parser.add_argument('source', help='Входной файл или папка')
    parser.add_argument('--bitrate', type=int, default=192, help='Битрейт (32-320 кбит/с)')
    parser.add_argument('--channels', type=int, choices=[1, 2], help='Количество каналов (1=моно, 2=стерео)')
    parser.add_argument('--sample-rate', type=int, help='Частота дискретизации (кГц)')
    parser.add_argument('--normalize', action='store_true', help='Нормализация громкости')
    parser.add_argument('--format', choices=SUPPORTED_OUTPUT, default='mp3', help='Выходной формат')
    parser.add_argument('--output', '-o', help='Выходной файл или папка')
    parser.add_argument('--recursive', action='store_true', help='Рекурсивный обход папок')
    parser.add_argument('--formats', action='store_true', help='Показать поддерживаемые форматы')
    args = parser.parse_args()

    if args.formats:
        print("Поддерживаемые входные форматы:", ', '.join(SUPPORTED_INPUT))
        print("Поддерживаемые выходные форматы:", ', '.join(SUPPORTED_OUTPUT))
        return

    check_ffmpeg()

    # Если указан выходной файл, а вход — папка — ошибка
    if args.output and os.path.isdir(args.source):
        print("Ошибка: при указании папки источником нельзя указать один выходной файл.", file=sys.stderr)
        sys.exit(1)

    # Поиск файлов
    files = find_audio_files(args.source, args.recursive)
    if not files:
        print(f"Не найдено аудиофайлов в {args.source}")
        sys.exit(1)

    # Подготовка выходной директории
    output_path = Path(args.output) if args.output else None
    if output_path and output_path.suffix:
        # Это конкретный файл
        if len(files) > 1:
            print("Ошибка: указан один выходной файл, но найдено несколько входных.", file=sys.stderr)
            sys.exit(1)
        output_path.parent.mkdir(parents=True, exist_ok=True)
    else:
        # Это папка
        out_dir = output_path if output_path else Path("./converted")
        out_dir.mkdir(parents=True, exist_ok=True)

    total = len(files)
    print(f"Найдено {total} аудиофайлов.")
    success = 0

    for i, input_file in enumerate(files, 1):
        if output_path and output_path.suffix:
            # Одиночный файл
            out_file = output_path
        else:
            # Сохраняем в папку с сохранением структуры
            rel = input_file.relative_to(args.source) if Path(args.source).is_dir() else input_file.name
            ext = f".{args.format}"
            out_file = (out_dir if output_path else Path("./converted")) / rel.with_suffix(ext)
            out_file.parent.mkdir(parents=True, exist_ok=True)

        print(f"[{i}/{total}] Конвертация {input_file} -> {out_file}")
        if convert_file(str(input_file), str(out_file), args.bitrate, args.channels,
                        args.sample_rate, args.normalize, args.format):
            success += 1
        else:
            print(f"  Ошибка при конвертации {input_file}")

    print(f"Готово! Успешно: {success}, Всего: {total}")

if __name__ == '__main__':
    main()

File: mp3_convert.rs
// mp3_convert.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::env;
use std::sync::Arc;
use std::sync::mpsc;
use std::thread;
use clap::{Parser, Arg};

#[derive(Parser)]
#[command(name = "mp3_convert")]
struct Args {
    #[arg(help = "Входной файл или папка")]
    source: String,
    #[arg(short, long, default_value_t = 192)]
    bitrate: u32,
    #[arg(short = 'c', long)]
    channels: Option<u8>,
    #[arg(short = 'r', long)]
    sample_rate: Option<u32>,
    #[arg(long)]
    normalize: bool,
    #[arg(short, long, default_value = "mp3")]
    format: String,
    #[arg(short = 'o', long)]
    output: Option<String>,
    #[arg(long)]
    recursive: bool,
    #[arg(long)]
    formats: bool,
}

const SUPPORTED_INPUT: &[&str] = &[".wav", ".flac", ".ogg", ".m4a", ".aac", ".wma", ".aiff", ".mp3", ".m4b"];
const SUPPORTED_OUTPUT: &[&str] = &["mp3", "wav", "flac", "ogg", "m4a", "aac"];

fn check_ffmpeg() {
    if Command::new("ffmpeg").arg("-version").output().is_err() {
        eprintln!("Ошибка: ffmpeg не найден. Установите FFmpeg и добавьте в PATH.");
        std::process::exit(1);
    }
}

fn build_ffmpeg_cmd(input: &str, output: &str, bitrate: u32, channels: Option<u8>,
                    sample_rate: Option<u32>, normalize: bool, format: &str) -> Vec<String> {
    let mut cmd = vec!["ffmpeg".to_string(), "-i".to_string(), input.to_string(), "-y".to_string()];
    cmd.push("-b:a".to_string());
    cmd.push(format!("{}k", bitrate));
    if let Some(ch) = channels {
        cmd.push("-ac".to_string());
        cmd.push(ch.to_string());
    }
    if let Some(sr) = sample_rate {
        cmd.push("-ar".to_string());
        cmd.push(sr.to_string());
    }
    if normalize {
        cmd.push("-af".to_string());
        cmd.push("loudnorm=I=-16:LRA=11:TP=-1.5".to_string());
    }
    let codec = match format {
        "mp3" => "libmp3lame",
        "wav" => "pcm_s16le",
        "flac" => "flac",
        "ogg" => "libvorbis",
        "m4a" | "aac" => "aac",
        _ => "libmp3lame",
    };
    cmd.push("-c:a".to_string());
    cmd.push(codec.to_string());
    cmd.push(output.to_string());
    cmd
}

fn convert_file(input: &str, output: &str, bitrate: u32, channels: Option<u8>,
                sample_rate: Option<u32>, normalize: bool, format: &str) -> bool {
    let cmd = build_ffmpeg_cmd(input, output, bitrate, channels, sample_rate, normalize, format);
    println!("  Выполняется: {}", cmd.join(" "));
    match Command::new(&cmd[0]).args(&cmd[1..]).status() {
        Ok(status) => status.success(),
        Err(_) => false,
    }
}

fn find_audio_files(root: &str, recursive: bool) -> Vec<PathBuf> {
    let mut files = Vec::new();
    let path = Path::new(root);
    if path.is_file() {
        if let Some(ext) = path.extension() {
            if SUPPORTED_INPUT.contains(&ext.to_str().unwrap_or("").to_lowercase().as_str()) {
                files.push(path.to_path_buf());
            }
        }
        return files;
    }
    if !path.is_dir() {
        return files;
    }
    if recursive {
        for entry in walkdir::WalkDir::new(path) {
            if let Ok(entry) = entry {
                if entry.file_type().is_file() {
                    if let Some(ext) = entry.path().extension() {
                        let ext_lower = ext.to_str().unwrap_or("").to_lowercase();
                        if SUPPORTED_INPUT.contains(&ext_lower.as_str()) {
                            files.push(entry.path().to_path_buf());
                        }
                    }
                }
            }
        }
    } else {
        if let Ok(entries) = fs::read_dir(path) {
            for entry in entries.flatten() {
                let p = entry.path();
                if p.is_file() {
                    if let Some(ext) = p.extension() {
                        let ext_lower = ext.to_str().unwrap_or("").to_lowercase();
                        if SUPPORTED_INPUT.contains(&ext_lower.as_str()) {
                            files.push(p);
                        }
                    }
                }
            }
        }
    }
    files
}

fn main() {
    let args = Args::parse();

    if args.formats {
        println!("Поддерживаемые входные форматы: {}", SUPPORTED_INPUT.join(", "));
        println!("Поддерживаемые выходные форматы: {}", SUPPORTED_OUTPUT.join(", "));
        return;
    }

    check_ffmpeg();

    if !SUPPORTED_OUTPUT.contains(&args.format.as_str()) {
        eprintln!("Неизвестный формат: {}. Доступны: {}", args.format, SUPPORTED_OUTPUT.join(", "));
        std::process::exit(1);
    }

    let files = find_audio_files(&args.source, args.recursive);
    if files.is_empty() {
        println!("Не найдено аудиофайлов в {}", args.source);
        std::process::exit(1);
    }

    let out_dir = match args.output {
        Some(ref o) => {
            if Path::new(o).is_dir() {
                o.clone()
            } else {
                Path::new(o).parent().unwrap_or(Path::new(".")).to_str().unwrap().to_string()
            }
        }
        None => "./converted".to_string(),
    };
    fs::create_dir_all(&out_dir).unwrap();

    let total = files.len();
    println!("Найдено {} аудиофайлов.", total);
    let (tx, rx) = mpsc::channel();
    let threads = 4;
    let mut handles = vec![];

    let args = Arc::new(args);

    for chunk in files.chunks((total + threads - 1) / threads) {
        let chunk = chunk.to_vec();
        let tx = tx.clone();
        let out_dir = out_dir.clone();
        let args = args.clone();
        handles.push(thread::spawn(move || {
            for (i, input_file) in chunk.iter().enumerate() {
                let rel = input_file.strip_prefix(&args.source).unwrap_or(input_file);
                let out_path = Path::new(&out_dir).join(rel.with_extension(&args.format));
                if let Some(parent) = out_path.parent() {
                    fs::create_dir_all(parent).unwrap();
                }
                tx.send((i+1, format!("Конвертация {} -> {}", input_file.display(), out_path.display()))).unwrap();
                let success = convert_file(
                    input_file.to_str().unwrap(),
                    out_path.to_str().unwrap(),
                    args.bitrate,
                    args.channels,
                    args.sample_rate,
                    args.normalize,
                    &args.format,
                );
                if !success {
                    tx.send((i+1, format!("  Ошибка при конвертации {}", input_file.display()))).unwrap();
                }
            }
        }));
    }
    drop(tx);
    let mut count = 0;
    for (idx, msg) in rx {
        count += 1;
        println!("[{}/{}] {}", count, total, msg);
    }
    for h in handles {
        h.join().unwrap();
    }
    println!("Готово!");
}
